### PR TITLE
[#IOCIT-182] add fn public

### PR DESCRIPTION
**Pull request by arcogabbo**
### List of changes
* created `function_public` with azure functions runtime v4
* created `shared_plan_1` service plan

### Motivation and context
porting from old infrastructure and various improvements

### Type of changes

- [x] Add new resources
- [ ] Update configuration to existing resources
- [ ] Remove existing resources

### Env to apply

- [ ] DEV
- [ ] UAT
- [x] PROD

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [ ] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [ ] No

### Other information

---

### How to apply

After PR is approved
1. run deploy pipeline from Azure DevOps [io-platform-iac-projects](https://dev.azure.com/pagopaspa/io-platform-iac-projects/_build?view=folders)
2. select PR branch
3. wait for approval

### CODE REVIEW

File: src/core/99_variables.tf [hunk 1]
```diff
@@ -790,3 +795,46 @@ variable "function_cgn_autoscale_default" {
...
 }
+
+# Shared functions
+variable "plan_shared_kind" {
+  type        = string
+  description = "App service plan kind"
+  default     = null
+}
+
+variable "plan_shared_sku_tier" {
```
Comment (pasqualedevita) [inline src/core/99_variables.tf:806]:
```suggestion
variable "plan_shared_1_sku_tier" {
```
add 1 to respect naming convention

File: src/core/99_variables.tf [hunk 2]
```diff
@@ -222,6 +222,11 @@ variable "cidr_subnet_cgn" {
   description = "Function cgn address space."
 }
 
+variable "cidr_subnet_shared" {
```
Comment (pasqualedevita) [inline src/core/99_variables.tf:225]:
```suggestion
variable "cidr_subnet_shared_1" {
```

File: src/core/function_public.tf [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+      FETCH_KEEPALIVE_TIMEOUT             = "60000"
+
+      COSMOSDB_URI      = data.azurerm_cosmosdb_account.cosmos_api.endpoint
+      COSMOSDB_KEY      = data.azurerm_cosmosdb_account.cosmos_api.primary_master_key
+      COSMOSDB_NAME     = "db"
+      StorageConnection = data.azurerm_storage_account.api.primary_connection_string
+
+      VALIDATION_CALLBACK_URL = "https://api-app.io.pagopa.it/email_verification.html"
+
+      SLOT_TASK_HUBNAME = "ProductionTaskHub"
```
Comment (pasqualedevita) [inline src/core/function_public.tf:31]:
```suggestion
```
managed by module

File: src/core/function_shared.tf [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    tier     = var.plan_shared_sku_tier
+    size     = var.plan_shared_sku_size
+    capacity = var.plan_shared_sku_capacity
+  }
+
+  tags = var.tags
+}
+
+# Subnet to host app function
+module "shared_snet" {
```
Comment (pasqualedevita) [inline src/core/function_shared.tf:24]:
```suggestion
module "shared_1_snet" {
```

File: src/core/function_shared.tf [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  source                                         = "git::https://github.com/pagopa/azurerm.git//subnet?ref=v1.0.51"
+  name                                           = format("%s-shared-snet", local.project)
```
Comment (pasqualedevita) [inline src/core/function_shared.tf:26]:
```suggestion
  name                                           = format("%s-shared-1-snet", local.project)
```

File: src/core/function_shared.tf [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+#
+# Function public
+#
+
+data "azurerm_storage_account" "iopstfn3public" {
+  name                = "iopstfn3public"
+  resource_group_name = azurerm_resource_group.rg_internal.name
+}
```
Comment (pasqualedevita) [inline src/core/function_shared.tf:54]: why this? what contains this storage?

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: src/core/99_variables.tf (hunk 1 above)
Comment (gunzip) [reply src/core/99_variables.tf:806]: hi @pasqualedevita! Can you direct us to the documentation that contains the proper naming scheme? We are trying to be as independent as we can (this knowlede could prevent you some saturday work ;-)

File: src/core/function_shared.tf (hunk 5 above)
Comment (arcogabbo) [reply src/core/function_shared.tf]: solved in https://github.com/pagopa/io-infra/pull/316/commits/46bd3249e99fe4ac8b3562c8c766c5af51e804da

File: src/core/99_variables.tf (hunk 1 above)
Comment (arcogabbo) [reply src/core/99_variables.tf]: solved in https://github.com/pagopa/io-infra/pull/316/commits/46bd3249e99fe4ac8b3562c8c766c5af51e804da

File: src/core/function_public.tf (hunk 3 above)
Comment (arcogabbo) [reply src/core/function_public.tf]: solved in https://github.com/pagopa/io-infra/pull/316/commits/e4609ce320c364770dc54708c07bbf87d24b9346

File: src/core/function_shared.tf (hunk 6 above)
Comment (arcogabbo) [reply src/core/function_shared.tf]: i believe that reference is unnecessary. solved in https://github.com/pagopa/io-infra/pull/316/commits/13cc1a1218a72999b1a21033d6b6e41ddb4e0248

File: src/core/env/prod/terraform.tfvars [hunk 7]
```diff
@@ -30,6 +30,7 @@ cidr_subnet_fnelt                          = ["10.0.11.0/24"]
 cidr_subnet_fnpblevtdispatcher             = ["10.0.12.0/24"]
 cidr_subnet_appgateway                     = ["10.0.13.0/24"]
 cidr_subnet_redis_apim                     = ["10.0.14.0/24"]
+cidr_subnet_shared_1                       = ["10.0.16.0/26"]
```
Comment (arcogabbo) [inline src/core/env/prod/terraform.tfvars:33]: NOTE: .15 is being proposed for `functions-admin` in https://github.com/pagopa/io-infra/pull/315

File: src/core/function_public.tf [hunk 8]
```diff
@@ -0,0 +1,245 @@
+#
+# APP CONFIGURATION
+#
+
+locals {
+  function_public = {
+    app_settings_common = {
+      FUNCTIONS_WORKER_RUNTIME       = "node"
+      WEBSITE_NODE_DEFAULT_VERSION   = "14.16.0"
```
Comment (michaeldisaro) [inline src/core/function_public.tf:9]: this can be removed

**Review by michaeldisaro — APPROVED** [review]

File: src/core/function_public.tf [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+
+module "function_public_staging_slot" {
+  source = "git::https://github.com/pagopa/azurerm.git//function_app_slot?ref=v3.4.0"
+
+  name                = "staging"
+  location            = var.location
+  resource_group_name = azurerm_resource_group.shared_rg.name
+  function_app_name   = module.function_public.name
+  function_app_id     = module.function_public.id
+  app_service_plan_id = azurerm_app_service_plan.shared_1_plan.id
```
Comment (pasqualedevita) [inline src/core/function_public.tf:83]:
```suggestion
  app_service_plan_id = azurerm_app_service_plan.shared_1_plan.id
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]
try this

**Comment by arcogabbo** [comment]
to address terraform resource attributes determination while performing plan, I will add the function in another PR and run the apply of the resources in this PR. therefore in [this](https://github.com/pagopa/io-infra/pull/316/commits/0b0c64dd7b3e09f588a4ec1225429bf2833edba4) commit, i removed the `function_public.tf` file

File: src/core/function_public.tf (hunk 9 above)
Comment (michaeldisaro) [reply src/core/function_public.tf]: what is the difference?

**Review by pasqualedevita — APPROVED** [review]

**Pull request merged**